scipy/optimize/_hessian_update_strategy.py

```diff
@@ -0,0 +1,307 @@
+"""Implementation of Quasi-Newton update strategies."""
+
+from __future__ import division, print_function, absolute_import
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import get_blas_funcs
+
+
+class HessianUpdateStrategy(object):
+    """Quasi-Newton Hessian update strategy."""
+
+    _syr = get_blas_funcs('syr', dtype='d')  # Symetric rank 1 update
+    _syr2 = get_blas_funcs('syr2', dtype='d')  # Symetric rank 2 update
+    _symv = get_blas_funcs('symv', dtype='d')  # Symetric matrix-vector product
+
+    def __init__(self, init_scale='auto'):
+        self.init_scale = init_scale
+        self.first_iteration = True
+
+    def initialize(self, n, approx_type):
+        """Initialize internal matrix.
+
+        Alocate internal memory for storing and updating
+        the Hessian or its inverse.
+
+        Parameters
+        ----------
+        n : int
+            Problem dimension.
+        approx_type : {'hess', 'inv_hess'}
+            Selects either the Hessian or the inverse Hessian.
+            When set to 'hess' the Hessian will be stored and updated.
+            When set to 'inv_hess' its inverse will be used instead.
+        """
+        self.approx_type = approx_type
+        if approx_type not in ('hess', 'inv_hess'):
+            raise ValueError("Unknown approx_type.")
+        # Create matrix
+        if self.approx_type == 'hess':
+            self.B = np.eye(n, dtype=float)
+        else:
+            self.H = np.eye(n, dtype=float)
+
+    def update(self, delta_x, delta_grad):
+        """Update matrix.
+
+        Update Hessian matrix or its inverse (depending on how 'approx_type'
+        is defined) using information about the last evaluated points.
+
+        Parameters
+        ----------
+        delta_x : ndarray
+            The difference between two points the gradient
+            function have been evaluated at: ``delta_x = x2 - x1``.
+        delta_grad : ndarray
+            The difference between the gradients:
+            ``delta_grad = grad(x2) - grad(x1)``.
+        """
+        raise NotImplementedError("The method ``update(delta_x, delta_grad)``"
+                                  " is not implemented.")
+
+    def _scale_matrix(self, delta_x, delta_grad):
+        # Scale matrix with user specified constant
+        try:
```

Comment:
I don't really like this code. If you restrict `self.init_scale` to 'auto' or float, why don't you write it as `if self.init_scale == 'auto'`.

Another question: should we really restrict it to those? Or maybe it's better to remove any scaling strategies from the base class? You could implement `_auto_scale` as a helper method in the base class, and the derived classes might choose to use it or not.

Reply:
I agree with you that it would be too restrictive to allow ``self.init_scale`` to be set only to to 'auto' or float. But now that I have created another level of abstraction (read the response to the above comment), I think it is not much of a problem.